elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java

```diff
@@ -0,0 +1,335 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.elasticsearch.util.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.nio.ContentEncoder;
+import org.apache.http.nio.IOControl;
+import org.apache.http.nio.entity.HttpAsyncContentProducer;
+import org.apache.http.protocol.HTTP;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+/**
+ * Optimised adapter to encode GSON objects into HttpEntity instances.
+ * The naive approach was using various StringBuilders; the objects we
+ * need to serialise into JSON might get large and this was causing the
+ * internal StringBuilder buffers to need frequent resizing and cause
+ * problems with excessive allocations.
+ *
+ * Rather than trying to guess reasonable default sizes for these buffers,
+ * we can defer the serialisation to write directly into the ByteBuffer
+ * of the HTTP client, this has the additional benefit of making the
+ * intermediary buffers short lived.
+ *
+ * The one complexity to watch for is flow control: when writing into
+ * the output buffer chances are that not all bytes are accepted; in
+ * this case we have to hold on the remaining portion of data to
+ * be written when the flow control is re-enabled.
+ *
+ * A side effect of this strategy is that the total content length which
+ * is being produced is not known in advance.
+ *
+ * @author Sanne Grinovero (C) 2017 Red Hat Inc.
+ */
+public final class GsonHttpEntity implements HttpEntity, HttpAsyncContentProducer, DigestSelfSigningCapable {
+
+	private static byte[] NEWLINE = "\n".getBytes( StandardCharsets.UTF_8 );
+
+	private static final BasicHeader CONTENT_TYPE = new BasicHeader( HTTP.CONTENT_TYPE, ContentType.APPLICATION_JSON.toString() );
+
+	private final Gson gson;
+	private final List<JsonObject> bodyParts;
+
+	/**
+	 * We don't want to compute the length in advance as it would defeat all optimisations.
+	 * Still it's possible that we happen to find out, for example if a Digest from all
+	 * content needs to be computed.
```

Comment:
This was actually buggy and I figured it out later: once the content length is provided we need to stick to it or the http client driver gets seriously confused. The new design is better and guarantees a "stable" response w/o drawbacks.